ext/opcache/jit/zend_jit_ir.c

```diff
@@ -16685,19 +16763,125 @@ static bool zend_jit_may_be_in_reg(const zend_op_array *op_array, zend_ssa *ssa,
 	return 1;
 }
 
-static void jit_frameless_icall0(zend_jit_ctx *jit, const zend_op *opline)
+static ir_ref jit_frameless_observer(zend_jit_ctx *jit, int checked_stack, const zend_op *opline, ir_ref op1_ref, uint32_t op1_info, ir_ref op2_ref, uint32_t op2_info, ir_ref op1_data_ref, uint32_t op1_data_info, ir_ref res_ref) {
+	// JIT: zend_observer_handler_is_unobserved(ZEND_OBSERVER_DATA(fbc))
+	ir_ref observer_handler;
+	zend_function *fbc = ZEND_FLF_FUNC(opline);
+	// Not need for runtime cache or generator checks here, we just need if_unobserved
+	ir_ref if_unobserved = jit_observer_fcall_is_unobserved_start(jit, fbc, &observer_handler, IR_UNUSED, IR_UNUSED).if_unobserved;
+
+	// push args to a valid call frame
+	bool track_last_valid_opline = jit->track_last_valid_opline;
+	bool use_last_valid_opline = jit->use_last_valid_opline;
+	const zend_op *last_valid_opline = jit->last_valid_opline;
+	bool reuse_ip = jit->reuse_ip;
+	// zend_jit_save_call_chain, but preserve delayed_call_level. We cannot override delayed_call_level without also saving in unobserved path.
+	if (jit->delayed_call_level) {
+		ir_ref call = jit_CALL(jit_FP(jit), call);
+		ir_STORE(jit_CALL(jit_IP(jit), prev_execute_data), jit->delayed_call_level == 1 ? IR_NULL : ir_LOAD_A(call));
+		ir_STORE(call, jit_IP(jit));
+	}
```

Comment:
You probably  need to call ``zend_jit_save_call_chain()`` 
I don't understand if there is any difference without ``JIT:`` comments.

Reply:
The problem is, as noted in the comment just above, that I cannot save the call chain conditionally:
```
// zend_jit_save_call_chain, but preserve delayed_call_level. We cannot override delayed_call_level without also saving in unobserved path.
```

Reply:
So no, there's no fundamental difference, just that we cannot just use that API.

Reply:
Instead of duplicating JIT code, it would be better to use the specially designed function and save and restore ``jit->delayed_call_level``.

Reply:
I can do that :-)